### PR TITLE
Feat/forgotten password

### PR DESCRIPTION
**Pull request by tnodell**
https://app.clickup.com/t/2704425/ID-5974

### CODE REVIEW

File: src/examples/rest/src/backend/schema/auth/forgotten-password/entity.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import {
```
Comment (tnodell) [inline src/examples/rest/src/backend/schema/auth/forgotten-password/entity.ts:1]: Why do we have so many "entities" (data or ORM) for forgotten password?/ Any authentication entity.

There's one here "src/examples/rest/src/backend/schema/auth/forgotten-password/entity.ts"

There's the Authentication entity in the MySQL db.  This Forgotten Password is one of these Authentications.
"src/backend/entities/mysql/authentication.ts"

There's another Authentication entity in:
"src/packages/auth/src/authentication/entities/authentication.ts"

Is one of these not needed?

File: src/examples/rest/src/backend/schema/auth/password/entity.ts [hunk 2]
```diff
@@ -5,6 +5,7 @@ import {
 } from '@exogee/graphweaver-auth';
 import { Credential as OrmCredential } from '../../../entities';
 
+// I don't think this Credential type is correct
```
Comment (tnodell) [inline src/examples/rest/src/backend/schema/auth/password/entity.ts:8]: This is grabbing the [Web API Credential](https://developer.mozilla.org/en-US/docs/Web/API/Credential), not sure how to type this.

File: src/packages/auth/src/authentication/resolvers/forgotten-password-link/base-resolver.ts [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+				throw new AuthenticationError('Authentication Failed: Link expired');
+			}
+
+			console.log('*******************\n');
+			console.log('link', link);
+			console.log('*******************\n');
+
+			// Get the user
+			const userProvider = EntityMetadataMap.get('User')?.provider as MikroBackendProvider<
+				UserProfile,
```
Comment (tnodell) [inline src/packages/auth/src/authentication/resolvers/forgotten-password-link/base-resolver.ts:151]: How do i type this?

File: src/packages/auth/src/authentication/resolvers/forgotten-password-link/base-resolver.ts [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+			const user = await userProvider.findOne({ id: link.userId });
+
+			if (!user?.id) {
+				throw new AuthenticationError('Authentication Failed: User not found');
+			}
+
+			// Get the user's credential
+
+			const credentialProvider = EntityMetadataMap.get('Credential')
+				?.provider as MikroBackendProvider<any, any>;
```
Comment (tnodell) [inline src/packages/auth/src/authentication/resolvers/forgotten-password-link/base-resolver.ts:164]: How do I type this?

File: src/packages/auth/src/authentication/resolvers/forgotten-password-link/base-resolver.ts [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+			console.log('*******************\n');
+			console.log('updatedCredential', updatedCredential);
+			console.log('*******************\n');
+
+			// redeem the link's token
+			const updatedLink = await provider.updateOne(link.id, {
+				data: {
+					...link.data,
+					redeemedAt: new Date(),
+				} as any,
```
Comment (tnodell) [inline src/packages/auth/src/authentication/resolvers/forgotten-password-link/base-resolver.ts:194]: type?

File: src/examples/rest/src/backend/schema/auth/forgotten-password/resolver.ts [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+		return true;
+	}
+
+	/**
+	 *
+	 * @param username fetch user details using a username
+	 * @returns return a UserProfile compatible entity
+	 */
+	async getUser(username: string): Promise<UserProfile> {
+		//@todo - extend the base loaders to handle filters - find by filter.
```
Comment (thekevinbrown) [inline src/examples/rest/src/backend/schema/auth/forgotten-password/resolver.ts:42]: Why is this needed? If it is needed, should we do it as part of this PR or remove the @todo? I'm pretty sure the base loaders can already find with a filter.

File: src/examples/rest/src/backend/schema/auth/forgotten-password/resolver.ts [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+		const provider = graphweaverMetadata.getEntity('Credential')?.provider;
+
+		if (!provider) throw new Error('Bad Request: Unknown provider.');
```
Comment (thekevinbrown) [inline src/examples/rest/src/backend/schema/auth/forgotten-password/resolver.ts:45]:
```suggestion
		if (!provider) throw new Error('Bad Request: No provider associated with the Credential entity.');
```

File: src/examples/rest/src/backend/schema/auth/forgotten-password/resolver.ts [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+		const user = await provider?.findOne({ username } as Filter<Credential<OrmCredential>>);
```
Comment (thekevinbrown) [inline src/examples/rest/src/backend/schema/auth/forgotten-password/resolver.ts:47]: Is the `as` necessary here, e.g.:
```suggestion
		const user = await provider?.findOne({ username });
```

File: src/packages/auth-ui-components/src/components/password/reset-password/component.tsx [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+		try {
+			const { data } = await resetPassword({
+				variables: {
+					password: values.password,
+					token,
+				},
+			});
+
+			const redirectUrl = redirectUri ?? new URL('/');
+			navigate(`/auth/login?redirect_uri=${redirectUrl}`, { replace: true });
```
Comment (thekevinbrown) [inline src/packages/auth-ui-components/src/components/password/reset-password/component.tsx:43]: Is this path guaranteed to be here?

File: src/packages/auth/src/authentication/resolvers/forgotten-password-link/base-resolver.ts [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+		@Mutation((returns) => Boolean)
+		async resetPassword(
+			@Arg('token', () => String) token: string,
+			@Arg('password', () => String) password: string
+		): Promise<boolean> {
+			const link = await this.getForgottenPasswordLink(token);
+
+			if (!link) {
+				logger.warn(`Failed to reset password: E0001: Link not found`);
+				throw new AuthenticationError('Reset Password Failed: E0001.');
```
Comment (thekevinbrown) [inline src/packages/auth/src/authentication/resolvers/forgotten-password-link/base-resolver.ts:128]: Since we're an open source project, people will be able to tell what these are from outside the solution. I reckon we log the full error but return a generic response.

File: src/packages/auth/src/authentication/resolvers/forgotten-password-link/base-resolver.ts [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+			}
+
+			if (link.createdAt < new Date(new Date().getTime() - ms(config.ttl))) {
+				logger.warn(`Failed to reset password: E0003: Link expired`);
+				throw new AuthenticationError('Reset Password Failed: E0003.');
+			}
+
+			// Get the user's credential
+			const credentialProvider = graphweaverMetadata.getEntity('Credential')
+				?.provider as BackendProvider<D, GraphQLEntity<D>>;
```
Comment (thekevinbrown) [inline src/packages/auth/src/authentication/resolvers/forgotten-password-link/base-resolver.ts:143]: Pretty sure the typecast is here to prevent the next line complaining about the provider being undefined, but it could be, so we should handle that instead:
```suggestion
				?.provider;
```

File: src/packages/auth/src/authentication/resolvers/forgotten-password-link/base-resolver.ts [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+				assertPasswordStrength: this.assertPasswordStrength,
+				provider: credentialProvider,
+				id: link.userId,
+				password,
+			});
+
+			// redeem the link's token
+			await provider.updateOne(link.id, {
+				data: {
+					...link.data,
```
Comment (thekevinbrown) [inline src/packages/auth/src/authentication/resolvers/forgotten-password-link/base-resolver.ts:156]: Do we need to update all the data here or just the `redeemedAt`?
```suggestion
```

File: src/packages/auth/src/authentication/resolvers/forgotten-password-link/resolver.ts [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+		async getForgottenPasswordLinks(
+			userId: string,
+			period: Date
+		): Promise<AuthenticationBaseEntity<ForgottenPasswordLinkData>[]> {
+			const existingLinks = await this.provider.find({
+				type: AuthenticationType.ForgottenPasswordLink,
+				userId,
+				createdAt_gt: period,
+			} as Filter<Authentication<D>> & {
+				createdAt_gt: Date; // @todo - extend filter<G> to include
```
Comment (thekevinbrown) [inline src/packages/auth/src/authentication/resolvers/forgotten-password-link/resolver.ts:94]: Suspect we could do this type work pretty easily, no? Then we could get rid of this typecast.

File: src/packages/auth/src/authentication/resolvers/password/resolver.ts [hunk 14]
```diff
@@ -141,18 +110,15 @@ export const createPasswordAuthResolver = <D extends BaseDataEntity>(
...
+			const entity = await updatePasswordCredential({
+				assertPasswordStrength: this.assertPasswordStrength,
+				provider: this.provider,
+				id: item.id,
+				password: item.password,
+				username: item.username,
+				params,
 			});
 
-			const [entity] = await this.runAfterHooks(HookRegister.AFTER_UPDATE, params, [credential]);
```
Comment (thekevinbrown) [inline src/packages/auth/src/authentication/resolvers/password/resolver.ts:153]: Out of curiosity, does this run anyway? If so, what triggers it?

File: src/packages/end-to-end/src/__tests__/api/auth/password/reset.test.ts [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+describe('Forgotten Password flow', () => {
+	beforeAll(async () => {
+		await ConnectionManager.connect('InMemory', connection);
+		const database = ConnectionManager.database('InMemory');
+		await database?.orm.schema.createSchema();
+
+		const credential = new OrmCredential();
+		credential.username = 'test';
+		credential.password = 'forgotPassword';
+		database?.em.persistAndFlush(credential);
```
Comment (thekevinbrown) [inline src/packages/end-to-end/src/__tests__/api/auth/password/reset.test.ts:127]:
```suggestion
		await database?.em.persistAndFlush(credential);
```

File: src/packages/end-to-end/src/__tests__/api/auth/password/reset.test.ts [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+			`,
+			variables: {
+				token,
+				password: 'newPassword',
+			},
+		});
+
+		assert(resetPasswordResponse.body.kind === 'single');
+		expect(resetPasswordResponse.body.singleResult.errors).toBeUndefined();
+		expect(resetPasswordResponse.body.singleResult.data?.resetPassword).toBe(true);
```
Comment (thekevinbrown) [inline src/packages/end-to-end/src/__tests__/api/auth/password/reset.test.ts:164]: Can we add logging in after the reset with the new password to make sure it works?

**Review by thekevinbrown — CHANGES_REQUESTED** [review]
Some stuff to work through, thanks for getting this over the line!

File: src/packages/auth/src/authentication/resolvers/forgotten-password-link/base-resolver.ts (hunk 12 above)
Comment (steve-keep) [reply src/packages/auth/src/authentication/resolvers/forgotten-password-link/base-resolver.ts:155]: Removing this line will also remove `token` from the data json payload that saved back to the db?

File: src/packages/auth/src/authentication/resolvers/password/resolver.ts (hunk 14 above)
Comment (steve-keep) [reply src/packages/auth/src/authentication/resolvers/password/resolver.ts:153]: Might need to talk about this one.

Currently the base resolver runs the hooks, but here we are not using the base resolver so we need to call them directly.

File: src/packages/auth/src/authentication/resolvers/forgotten-password-link/base-resolver.ts (hunk 10 above)
Comment (steve-keep) [reply src/packages/auth/src/authentication/resolvers/forgotten-password-link/base-resolver.ts:128]: Can you provide an example of what you mean here then I can follow.

File: src/packages/auth-ui-components/src/components/password/reset-password/component.tsx (hunk 9 above)
Comment (steve-keep) [reply src/packages/auth-ui-components/src/components/password/reset-password/component.tsx:43]: Guaranteed how?

File: src/examples/rest/src/backend/schema/auth/forgotten-password/resolver.ts (hunk 8 above)
Comment (steve-keep) [reply src/examples/rest/src/backend/schema/auth/forgotten-password/resolver.ts:48]: The return type is `Filter<GraphQLEntity<any>>` so we would need to fix that to remove the cast.

File: src/examples/rest/src/backend/schema/auth/forgotten-password/resolver.ts (hunk 6 above)
Comment (steve-keep) [reply src/examples/rest/src/backend/schema/auth/forgotten-password/resolver.ts:42]: No our Base loaders only take an ID and here we have a username. 

I think we can remove the todo in this case as we are no using the base loader but the provider directly.

Comment (thekevinbrown) [reply src/examples/rest/src/backend/schema/auth/forgotten-password/resolver.ts]: Yup, sounds good!

File: src/packages/auth-ui-components/src/components/password/reset-password/component.tsx (hunk 9 above)
Comment (thekevinbrown) [reply src/packages/auth-ui-components/src/components/password/reset-password/component.tsx:43]: As in is it possible that the developer sets their login page up at `/log-me-in` instead of `/auth/login`?

File: src/packages/auth/src/authentication/resolvers/forgotten-password-link/base-resolver.ts (hunk 10 above)
Comment (thekevinbrown) [reply src/packages/auth/src/authentication/resolvers/forgotten-password-link/base-resolver.ts:128]:
```suggestion
				throw new AuthenticationError('Reset Password Failed');
```

File: src/packages/auth/src/authentication/resolvers/password/resolver.ts (hunk 14 above)
Comment (steve-keep) [reply src/packages/auth/src/authentication/resolvers/password/resolver.ts:153]: This was moved to the utils folder

File: src/packages/auth-ui-components/src/components/password/reset-password/component.tsx (hunk 9 above)
Comment (steve-keep) [reply src/packages/auth-ui-components/src/components/password/reset-password/component.tsx:43]: ah I see what you mean!

No is the answer 🤔

File: src/packages/auth/src/authentication/resolvers/forgotten-password-link/resolver.ts (hunk 13 above)
Comment (steve-keep) [reply src/packages/auth/src/authentication/resolvers/forgotten-password-link/resolver.ts]: Removed now operations are apart of the filter type

File: src/packages/auth-ui-components/src/components/password/reset-password/component.tsx (hunk 9 above)
Comment (steve-keep) [reply src/packages/auth-ui-components/src/components/password/reset-password/component.tsx]: Moved the values to env vars 👍

File: src/examples/rest/src/backend/schema/auth/forgotten-password/resolver.ts (hunk 8 above)
Comment (steve-keep) [reply src/examples/rest/src/backend/schema/auth/forgotten-password/resolver.ts]: The provider is now correctly typed.

**Review by thekevinbrown — APPROVED** [review]
